sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala

```diff
@@ -145,11 +145,19 @@ case class Not(child: Expression)
   override def sql: String = s"(NOT ${child.sql})"
 }
 
-/**
- * Evaluates to `true` if `values` are returned in `query`'s result set.
- */
-case class InSubquery(values: Seq[Expression], query: ListQuery)
-  extends Predicate with Unevaluable {
+abstract class PredicateSubquery extends Predicate with Unevaluable {
+
+  def values: Seq[Expression]
+  def query: ListQuery
+  def genCmp: (Expression, Expression) => Expression
+  def symbol: String
+
+  @transient protected lazy val cmpSymbol: String = genCmp(value, query) match {
+    case b: BinaryComparison =>
+      b.symbol
+    case _ @ Not(EqualTo(_, _)) =>
```

Comment:
We must handle NOT in the PredicateSubquery node? We cannot handle like `Not(PredicateSubquery)`?

Reply:
I think so... These are two different cases. We just need to extract the comparison symbol here, whose SQL should be "x != ANY (query)". But `Not(PredicateSubquery)` means "NOT x cmp ANY(query)".

Reply:
How about specially handling that case (Not(AnySubquery)) in `RewritePredicateSubquery`?

Reply:
As we have discussed, `Not(AnySubquery(EqualTo))` is different from `AnySubquery(Not(EqualTo))`. We don't need to specially handle `Not(AnySubquery)`.
We define `cmpSymbol` here, because we can't directly get the comparison symbol from an `Expression`. It is only called by `symbol`.

Reply:
For example, we can define `NotAnySubquery` for that special case and add `<>` for semi-join in `RewritePredicateSubquery`? (I'm not 100% sure about which is better, so it'd be better to wait for other developer's suggestions)